### PR TITLE
Move Frequencies, fftfreq, rfftfreq from DSP.jl to AbstractFFTs.jl

### LINKED ISSUE #19: Frequencies of FFT: `fftfreq`
I think it would make sense to add a `fftfreq` function here which returns the discrete FFT sample frequencies. Currently, a version is provided by DSP.jl [here](https://github.com/JuliaDSP/DSP.jl/blob/920480c90975d6cc7eb5b2e3e95f468dbf55e640/src/util.jl#L146). It seems odd that one has to add another package for getting the FFT frequencies.

Note that numpy [also has this](https://docs.scipy.org/doc/numpy/reference/generated/numpy.fft.fftfreq.html) as part of `numpy.fft`.

I could make a PR. What do you think?

### PR DESCRIPTION
**Pull request by crstnbr**
See discussions in #19 and https://github.com/JuliaDSP/DSP.jl/issues/298.

I have basically copy pasted the relevant code from DSP.jl and modified it according to @stevengj's [comment](https://github.com/JuliaDSP/DSP.jl/issues/298#issuecomment-491080346).

This way one can use, for example, a [measurement](https://github.com/JuliaPhysics/Measurements.jl) as sampling frequency:

```julia
julia> m = 5 ± 1
5.0 ± 1.0

julia> fftfreq(10, m)
10-element Frequencies{Measurement{Float64}}:
  0.0 ± 0.0
  0.5 ± 0.1
  1.0 ± 0.2
  1.5 ± 0.30000000000000004
  2.0 ± 0.4
 -2.5 ± 0.5
 -2.0 ± 0.4
 -1.5 ± 0.30000000000000004
 -1.0 ± 0.2
 -0.5 ± 0.1
```

Closes #19.

### CODE REVIEW

**Comment by crstnbr** [comment]
Note that a quantity from Unitful.jl doesn't currently work due to

```julia
julia> u = 3 * Unitful.Hz
3 Hz

julia> u isa Real
false
```

as has been indicated [here](https://github.com/JuliaDSP/DSP.jl/issues/298#issuecomment-491081477). Should we generalize to `<:Number` instead of `<:Real`?

File: src/AbstractFFTs.jl [hunk 1]
```diff
@@ -3,8 +3,8 @@ module AbstractFFTs
 export fft, ifft, bfft, fft!, ifft!, bfft!,
        plan_fft, plan_ifft, plan_bfft, plan_fft!, plan_ifft!, plan_bfft!,
        rfft, irfft, brfft, plan_rfft, plan_irfft, plan_brfft,
-       fftshift, ifftshift
+       fftshift, ifftshift, Frequencies, fftfreq, rfftfreq
 
 include("definitions.jl")
 
-end # module
+end # module
```
Comment (giordano) [inline src/AbstractFFTs.jl]: Missing newline here :slightly_smiling_face:

**Comment by crstnbr** [comment]
```julia
julia> using Unitful

julia> u = 3 * Unitful.Hz
3 Hz

julia> fftfreq(10, u)
10-element Frequencies{Quantity{Float64,�^-1,Unitful.FreeUnits{(Hz,),�^-1,nothing}}}:
                 0.0 Hz
                 0.3 Hz
                 0.6 Hz
  0.8999999999999999 Hz
                 1.2 Hz
                -1.5 Hz
                -1.2 Hz
 -0.8999999999999999 Hz
                -0.6 Hz
                -0.3 Hz
```

**Comment by giordano** [comment]
So now a number with both physical units and uncertainty should work, right? :smiley:

**Comment by crstnbr** [comment]
> So now a number with both physical units and uncertainty should work, right? 😃

Yes.

**Comment by crstnbr** [comment]
Maybe the optional sampling frequency should rather be a keyword argument (in the spirit of https://github.com/JuliaMath/AbstractFFTs.jl/issues/15)?

Otherwise this should be good to go.

File: src/definitions.jl [hunk 2]
```diff
@@ -396,6 +396,49 @@ function ifftshift(x,dim)
...
+struct Frequencies{T<:Number} <: AbstractVector{T}
+    nreal::Int
+    n::Int
+    multiplier::T
+end
+
+unsafe_getindex(x::Frequencies, i::Int) =
+    (i-1+ifelse(i <= x.nreal, 0, -x.n))*x.multiplier
+function Base.getindex(x::Frequencies, i::Int)
+    (i >= 1 && i <= x.n) || throw(BoundsError())
```
Comment (stevengj) [inline src/definitions.jl]: Could be
```jl
@boundscheck Base.checkbounds(x, i)
```
?

File: src/definitions.jl [hunk 3]
```diff
@@ -396,6 +396,49 @@ function ifftshift(x,dim)
...
+
+function Base.iterate(x::Frequencies, i::Int=1)
+    i > x.n ? nothing : (unsafe_getindex(x,i), i + 1)
+end
+Base.size(x::Frequencies) = (x.n,)
+Base.step(x::Frequencies) = x.multiplier
+
+"""
+    fftfreq(n, fs=1)
+Return discrete fourier transform sample frequencies. The returned
```
Comment (stevengj) [inline src/definitions.jl]: `Return the discrete Fourier transform (DFT) sample frequencies for a DFT of length `n`.

Missing backtick quotes on `Frequencies` etc.

File: src/definitions.jl [hunk 4]
```diff
@@ -396,6 +396,49 @@ function ifftshift(x,dim)
     circshift(x, s)
 end
 
+##############################################################################
+
+
+struct Frequencies{T<:Number} <: AbstractVector{T}
+    nreal::Int
+    n::Int
+    multiplier::T
```
Comment (stevengj) [inline src/definitions.jl]: The constructor should maybe have some sanity checks that `1 ≤ nreal ≤ n`?

File: src/definitions.jl [hunk 5]
```diff
@@ -396,6 +396,49 @@ function ifftshift(x,dim)
...
+end
+
+unsafe_getindex(x::Frequencies, i::Int) =
+    (i-1+ifelse(i <= x.nreal, 0, -x.n))*x.multiplier
```
Comment (stevengj) [inline src/definitions.jl]: This seems like it should be `i-1 ≤ x.nreal`

File: test/runtests.jl [hunk 6]
```diff
@@ -81,6 +81,23 @@ end
     @test AbstractFFTs.ifftshift([1 2 3; 4 5 6], 1:2) == [5 6 4; 2 3 1]
 end
 
+@testset "FFT Frequencies" begin
+    # N even
+    @test fftfreq(8) == [0.0, 0.125, 0.25, 0.375, -0.5, -0.375, -0.25, -0.125]
+    @test rfftfreq(8) == [0.0, 0.125, 0.25, 0.375, 0.5]
```
Comment (stevengj) [inline test/runtests.jl]: 0.5 should be positive in both cases to be consistent.

File: src/definitions.jl [hunk 7]
```diff
@@ -396,6 +396,49 @@ function ifftshift(x,dim)
...
+Base.step(x::Frequencies) = x.multiplier
+
+"""
+    fftfreq(n, fs=1)
+Return discrete fourier transform sample frequencies. The returned
+Frequencies object is an AbstractVector containing the frequency
+bin centers at every sample point. `fs` is the sample rate of the
+input signal.
+"""
+fftfreq(n::Int, fs::Number=1) = Frequencies(((n-1) >> 1)+1, n, fs/n)
```
Comment (stevengj) [inline src/definitions.jl]: `(n+1) >> 1` is equivalent and seems simpler.

File: src/definitions.jl (hunk 5 above)
Comment (stevengj) [reply src/definitions.jl]: Oh, I guess by `nreal` you mean the number of nonnegative frequencies, in which case `i ≤ x.nreal` is right.   But maybe call it `n_nonnegative` or something like that to be clearer?

File: src/definitions.jl (hunk 7 above)
Comment (stevengj) [reply src/definitions.jl]: If we want the first argument to be the number of nonnegative frequencies, then I think it should be `(n+3) >> 1`

**Comment by stevengj** [comment]
Oh, I see the reason for assigning the Nyquist element (the middle frequency for even `n`) to be a "negative" frequency — it is consistent with what `fftshift` does.

Of course, it's completely arbitrary whether we call this frequency positive or negative, but it's annoying that it is inconsistent with the real-input FFT case where we call all the frequencies nonnegative.    But I guess we can't change the `fftshift` function at this point.  😢

File: src/definitions.jl (hunk 4 above)
Comment (crstnbr) [reply src/definitions.jl]: Now we `error` if the condition isn't satisfied.

Is there a more specific error that we could throw instead? Maybe a `DomainError`?

Comment (stevengj) [reply src/definitions.jl]: Usually this would be `ArgumentError`

**Comment by crstnbr** [comment]
Ready for merge?

**Comment by galenlynch** [comment]
It might be worth thinking about how this change affects other packages. E.g. how will this affect DSP.jl, where this code was taken from, and do other packages depend on these functions being in DSP.jl? It would be great if you could also open PRs against other affected packages to accommodate this change.

**Comment by crstnbr** [comment]
> It might be worth thinking about how this change affects other packages.

Well, this PR is only adding a new feature. Therefore, technically, no other package is affected - the PR is trivially non-breaking. Maybe DSP.jl is an exception because I took the code from there and we probably want to avoid multiple identical definitions in different packages. I'll create a PR over there to switch to using AbstractFFTs `fftfreq` etc. in DSP.jl. However, technically this shouldn't affect any other packages either as this change will also be non-breaking (right?).

If you're suggesting to make PRs against all packages that use DSP.jl for the sole purpose of getting `fftfreq` and companions, this sounds like quite a bit of work, which I'm not sure is worth doing, since these packages will this work fine. How would I even find those packages? Any ideas?

**Comment by crstnbr** [comment]
DSP.jl PR created: https://github.com/JuliaDSP/DSP.jl/pull/300

However, it can only be incorporated after this PR here has been merged.

**Comment by galenlynch** [comment]
Maybe it would be enough to have DSP.jl still export `rfftfreq`, `fftfreq`, etc. but add some deprecation warning to let users know that its location will be changing?

**Comment by heitorPB** [comment]
Is it possible to have a new release of AbstractFFTs with this PR?

**Comment by stevengj** [comment]
Just need a PR to update Project.toml, I think?

**Comment by galenlynch** [comment]
@stevengj Could you please tag a new version so we can fix the consequences of this in DSP.jl (JuliaDSP/DSP.jl#306)?

**Pull request merged**